include/ginkgo/core/solver/ir.hpp

```diff
@@ -145,33 +139,15 @@ class Ir : public EnableLinOp<Ir<ValueType>>, public Transposable {
      *
      * @param new_solver  the new inner solver
      */
-    void set_solver(std::shared_ptr<const LinOp> new_solver)
-    {
-        GKO_ASSERT_EQUAL_DIMENSIONS(new_solver, this);
-        solver_ = new_solver;
-    }
+    void set_solver(std::shared_ptr<const LinOp> new_solver);
 
-    /**
-     * Gets the stopping criterion factory of the solver.
-     *
-     * @return the stopping criterion factory
-     */
-    std::shared_ptr<const stop::CriterionFactory> get_stop_criterion_factory()
-        const
-    {
-        return stop_criterion_factory_;
-    }
+    Ir& operator=(const Ir& other);
 
-    /**
-     * Sets the stopping criterion of the solver.
-     *
-     * @param other  the new stopping criterion factory
-     */
-    void set_stop_criterion_factory(
-        std::shared_ptr<const stop::CriterionFactory> other)
-    {
-        stop_criterion_factory_ = std::move(other);
-    }
+    Ir& operator=(Ir&& other);
+
+    Ir(const Ir& other);
+
+    Ir(Ir&& other);
```

Comment:
It might make sense to have a `IterativeSolverBase`/`EnableIterativeSolverBase` to move these copy and move assignment and constructors to those mixins.

Reply:
Ir is kind of a special case, since it has both a system matrix and a solver, so we need assignment operators specialized for them. For all the other solvers, that is true :)